### ISSUE TITLE
Is this project dead?

### ISSUE THREAD

**Issue by zdk123** [danoneata/fast-nca#1]
I came across your thesis/this repo via a search for NCA methods and I like your approach! Do you have plans to address any of the items in the WIP?

**Comment by danoneata** [comment]
Hello Zach!

Thanks for the interest in the project and for reminding me of its existence! 😃

In principle, I was not planning of continuing its development and I wonder how appropriate the current to-do list is. Back in the day, when I started working on this project (~2011), you had to carefully compute all the gradients by hand and make sure they are efficiently computed—this is the main reason of the convoluted code 😁 Nowadays, I would much rather rely on a library that supports automatic differentiation out of the box (such as PyTorch or TensorFlow); so if I were to revive the repository, the first thing I would do would be to clean up the implementation by delegating the gradient computation to PyTorch. That said, there are some recent implementations [[1](https://github.com/KevinMusgrave/pytorch-metric-learning), [2](https://github.com/microsoft/snca.pytorch), [3](https://github.com/dichotomies/proxy-nca)] that are very similar in spirit to these ideas.

What would features would you be interested to see the most in an existing implementation of NCA? Maybe we can get something together. :-)

**Comment by zdk123** [comment]
I might take some time to reformulate my thinking in an email. But your general claim that automatic differentiation > explicit computing a gradient is intriguing. I'm not that familiar with PyTorch et al - do you a citable reference for this claim (links look good though, thanks!)

**Comment by danoneata** [comment]
Some useful resources on automatic differentiation (AD) might be _(i)_ Justin Domke's [blog-post](https://justindomke.wordpress.com/2009/02/17/automatic-differentiation-the-most-criminally-underused-tool-in-the-potential-machine-learning-toolbox/), which was written in 2009, but provides some context on the use of AD in the machine learning community, and _(ii)_ a recent and thorough [survey on the topic from Baydin _et. al_](https://arxiv.org/pdf/1502.05767.pdf). I have learnt of automatic differentiation from Iain Murray, while working on my master's thesis under his supervision—you may find his [course notes](https://www.inf.ed.ac.uk/teaching/courses/mlpr/2019/notes/w7c_backprop.html) useful :slightly_smiling_face:
